### PR TITLE
config: persist default store limit for future stores

### LINKED ISSUE #10898: Persist default store limit for future stores
## Problem

When calling the PD API to update all store limits, for example:

```bash
curl -X POST http://<pd>/pd/api/v1/stores/limit -d '{"rate": 60, "type": "add-peer"}'
```

PD updates all existing stores' `add-peer` limit as expected. New TiKV stores may also inherit the new value while the same PD/API/scheduling process remains alive, because the process-local default store limit was updated.

However, after the relevant PD/API/scheduling process restarts or leadership/primary handling switches, newly joined TiKV stores can fall back to the built-in default, currently `add-peer: 15`, because the default value used for future stores is not persisted as cluster configuration.

## Expected behavior

After setting all stores' limit through `/pd/api/v1/stores/limit`, future TiKV stores should consistently inherit the configured default even after PD/API/scheduling service restart or leader/primary switch.

## Current behavior

- Existing stores get persisted per-store entries under `schedule.store-limit[storeID]`.
- Future stores are initialized from a process-local default store limit.
- That process-local default can reset after restart/switch, so future stores may inherit `15` instead of the previously set value.

## Proposed direction

Persist a schedule-level default store limit, for example:

```toml
[schedule.default-store-limit]
add-peer = 60
remove-peer = 15
```

Then use this priority when initializing or reading store limits:

1. per-store `schedule.store-limit[storeID]`
2. persisted `schedule.default-store-limit`
3. process default

The per-store override should remain highest priority.

### PR DESCRIPTION
**Pull request by King-Dylan**
### What problem does this PR solve?

Issue Number: close #10898

Persist the default store limit used for future stores. Today `/pd/api/v1/stores/limit` updates existing stores and mutates the process-local default store limit, but the default for future stores is not persisted. After PD/API/scheduling service restart or leader/primary switch, newly joined TiKV stores can fall back to the built-in default value.

### What is changed and how does it work?

- Add `schedule.default-store-limit` to `ScheduleConfig`.
- Update `SetAllStoresLimit` to persist the default limit as well as updating existing per-store entries.
- Initialize missing/new store limits from persisted `schedule.default-store-limit` before falling back to the process default.
- Keep per-store `schedule.store-limit[storeID]` as the highest-priority override.
- Keep the current TiFlash default store limit behavior unchanged.
- Apply the same default-store-limit logic to MCS scheduling `PersistConfig`.
- Preserve explicit zero values and migrate legacy `store-balance-rate` or old PD/MCS config responses without overwriting explicitly configured fields.
- Serialize full-config persistence, store-limit updates, and scheduler-list initialization through the same read-modify-publish-persist boundary, so acknowledged defaults cannot be overwritten by stale schedule snapshots.
- Apply partial schedule-config JSON updates to the latest clone inside the schedule persistence transaction, preserving unrelated concurrent fields without CAS-conflict 500 responses.
- Reject negative or non-finite default store limit values before persistence.

The resulting priority is:

1. per-store `schedule.store-limit[storeID]`
2. persisted `schedule.default-store-limit`
3. process default

During a mixed-version rolling upgrade, `/stores/limit` remains available and continues updating existing per-store limits without synchronously depending on every PD or Scheduling Service member. The persisted-default guarantee for future stores applies once every PD and Scheduling Service member is upgraded; if the API was used during the mixed-version window, reapply the desired default after the upgrade completes.

Downgrading to a pre-feature PD or Scheduling Service after changing the persisted default is not supported. A pre-feature binary does not understand the new field and may persist its built-in value for stores added while downgraded; that per-store value can remain effective after a later re-upgrade.

### Check List

Tests

- Unit test
- Integration test
- Manual E2E test:
[pd-pr-10900-e2e-test-report-en.md](https://github.com/user-attachments/files/31256605/pd-pr-10900-e2e-test-report-en.md)

> This report validates the core persistence behavior at commit `0bb1ff6e`: real PD leader transfer/restart, new TiKV store inheritance, and a live Scheduling Service watcher all retained `45/45`. Follow-up review commit `5c414acf7` keeps that core path and removes the reported CAS-conflict 500 behavior and the synchronous mixed-member preflight; those follow-ups are covered by targeted unit, race, HTTP, and MCS integration tests.

```bash
Main conclusion:

    - On master, the bug is reproducible:
      - After POST /pd/api/v1/stores/limit {"rate":45}, a store added before PD leader transfer gets 45/45.
      - After PD leader transfer, a newly added store falls back to the old default 15/15.

    - On PR #10900, the fix is verified:
      - A store added before PD leader transfer gets 45/45.
      - A store added after PD leader transfer also gets 45/45.
      - This confirms that the default store limit set via /pd/api/v1/stores/limit is now persisted into schedule.default-store-limit.

    Supplementary checks also passed:

    - PD restart preserves default-store-limit = 45/45.
    - Per-store override works and does not pollute the default value.
    - Invalid inputs such as rate=0, rate=-1, and invalid type return HTTP 400.
    - Legacy schedule config updates do not overwrite default-store-limit.
    - Concurrent updates may return CAS-conflict 500 responses for some schedule config requests, but the final default-store-limit remains correct at 45/45.

    MCS / scheduling service validation passed:

    - MCS store-limit and config watch tests passed.
    - A real scheduling service was started and successfully observed default-store-limit = 45/45.
    - Mixed-version compatibility gate and related API/config compatibility tests passed.

    Final assessment:

    PR #10900 achieves its intended fix. The persisted default store limit behaves correctly across PD leader transfer, PD restart, new store addition, MCS config watch, legacy config update, and compatibility-gate scenarios. Functional validation is successful.
```

The tests cover:

- explicit zero values and legacy `store-balance-rate` precedence;
- old PD and Scheduling Service config payload migration;
- persisted-default precedence after restart;
- `/stores/limit` remaining available while an independently registered Scheduling Service member is unavailable;
- serialization against concurrent full-config persistence, with the final value reloaded from storage;
- scheduler initialization merging only its scheduler list after a concurrent store-limit update, with both memory and reloaded storage verified;
- concurrent partial schedule-config updates merging into the latest config without CAS conflicts, with memory and persisted state compared;
- rejection of invalid persisted defaults.

Code changes

- Has configuration change

Side effects

- During a mixed-version rolling upgrade, `/stores/limit` remains available for existing stores, but future-store default persistence is guaranteed only after all PD and Scheduling Service members support the new field. Reapply the desired default after the upgrade if it was changed during the mixed-version window.
- After changing the persisted default, downgrade to a pre-feature PD or Scheduling Service is unsupported.

### Release note

```release-note
PD now persists store limit defaults set by the stores limit API, so new TiKV stores consistently inherit the configured default after leader switches or restarts. The existing stores-limit API remains available during rolling upgrades; future-store default persistence is guaranteed after all PD and Scheduling Service members support the new format. Downgrading to a version that predates the persisted format is unsupported after the default is changed.
```

### CODE REVIEW

**Review by bufferflies — APPROVED** [review]

File: pkg/mcs/scheduling/server/config/config.go [hunk 1]
```diff
@@ -517,10 +517,7 @@ func (o *PersistConfig) GetStoreLimit(storeID uint64) (returnSC sc.StoreLimitCon
 		return limit
 	}
 	cfg := o.GetScheduleConfig().Clone()
-	limitCfg := sc.StoreLimitConfig{
-		AddPeer:    sc.DefaultStoreLimit.GetDefaultStoreLimit(storelimit.AddPeer),
-		RemovePeer: sc.DefaultStoreLimit.GetDefaultStoreLimit(storelimit.RemovePeer),
-	}
+	limitCfg := cfg.GetDefaultStoreLimit()
```
Comment (rleungx) [inline pkg/mcs/scheduling/server/config/config.go:540]: Config entries written by pre-upgrade PD do not contain `default-store-limit`; the MCS watcher unmarshals those entries into a zero-valued `ScheduleConfig`, so this fallback installs `{0,0}` for an unlisted store. In store-limit v1, zero is treated as unlimited, so a new store whose per-store entry has not been persisted can bypass add/remove-peer throttling after an upgrade.

File: pkg/schedule/config/config.go [hunk 2]
```diff
@@ -489,6 +489,38 @@ func (c *ScheduleConfig) Adjust(meta *configutil.ConfigMetaData, reloading bool)
...
+}
+
+func (c *ScheduleConfig) migrateStoreBalanceRate(defaultStoreLimitMeta *configutil.ConfigMetaData) {
+	if c.StoreBalanceRate == 0 {
+		return
+	}
+	defaultStoreLimit := StoreLimitConfig{AddPeer: c.StoreBalanceRate, RemovePeer: c.StoreBalanceRate}
+	DefaultStoreLimit = StoreLimit{AddPeer: defaultStoreLimit.AddPeer, RemovePeer: defaultStoreLimit.RemovePeer}
+	if defaultStoreLimitMeta == nil {
+		if c.DefaultStoreLimit == (StoreLimitConfig{}) {
```
Comment (rleungx) [inline pkg/schedule/config/config.go]: During reload of a legacy config that still contains `store-balance-rate`, `Adjust(nil, true)` has already filled `DefaultStoreLimit` with 15 before JSON unmarshalling, so this zero-value check never migrates the legacy rate. The deprecated field is then cleared and future stores silently fall back to 15 instead of the persisted rate.

File: pkg/mcs/scheduling/server/config/config.go (hunk 1 above)
Comment (King-Dylan) [reply pkg/mcs/scheduling/server/config/config.go:540]: Fixed in 96d4b0317. Persisted JSON now tracks `default-store-limit` field presence, and the MCS watcher initializes and migrates the watched schedule config before installing it. Legacy configs without the field now use the built-in default (or `store-balance-rate` when present), while explicit zero values remain unchanged. `TestAdjustScheduleConfigDefaultStoreLimit` covers these upgrade cases and the future-store lookup.

File: pkg/schedule/config/config.go (hunk 2 above)
Comment (King-Dylan) [reply pkg/schedule/config/config.go]: Fixed in 96d4b0317. The persisted-config migration now uses per-field JSON presence instead of a zero-value sentinel, with precedence `default-store-limit` > `store-balance-rate` > process default. `TestReloadLegacyStoreBalanceRate` exercises the real `Adjust -> JSON unmarshal -> migrate` reload path and verifies that the legacy rate is retained for future stores.

File: pkg/schedule/config/config.go [hunk 3]
```diff
@@ -337,6 +343,34 @@ type ScheduleConfig struct {
...
+	var fields struct {
+		DefaultStoreLimit map[string]json.RawMessage `json:"default-store-limit"`
+	}
+	if err := json.Unmarshal(data, &fields); err != nil {
+		return err
+	}
+	_, addPeerDefined := fields.DefaultStoreLimit["add-peer"]
+	_, removePeerDefined := fields.DefaultStoreLimit["remove-peer"]
+	*c = ScheduleConfig(decoded)
+	c.defaultStoreLimitJSONPresence = &storeLimitConfigJSONPresence{
```
Comment (rleungx) [inline pkg/schedule/config/config.go]: When a new PD reads config from a pre-upgrade scheduling primary, that peer omits `default-store-limit`; this decoder leaves the field as `{0,0}`, and the HTTP bridge path does not run `MigrateDeprecatedFlags`. The existing full GET/modify/POST workflow then persists those zeros as explicit defaults, which store-limit v1 treats as unlimited for future stores.

Comment (King-Dylan) [reply pkg/schedule/config/config.go]: Fixed in d7a3c47a3. PD now runs `MigrateDeprecatedFlags` immediately after decoding full remote config from either the PD leader or scheduling primary, before the config can be serialized by the HTTP bridge. The scheduling-service follower merge path is normalized as well. `TestUnmarshalRemoteConfigMigratesDefaultStoreLimit` covers a missing field (process default), legacy `store-balance-rate`, explicit zero, and the JSON round trip; partial POST handling remains unchanged.

File: server/util.go [hunk 4]
```diff
@@ -58,6 +60,61 @@ func CheckPDVersionWithClusterVersion(opt *config.PersistOptions) {
...
+			"cannot update default store limit while %s member %s does not support persisted default store limits",
+			component, name)
+	}
+	return nil
+}
+
+// checkDefaultStoreLimitPersistenceSupport prevents a new persisted schedule
+// field from being activated while an old PD or Scheduling Service member can
+// still become leader/primary and drop the unknown field on its next persist.
+func (s *Server) checkDefaultStoreLimitPersistenceSupport() error {
```
Comment (rleungx) [inline server/util.go]: Once a default has been activated, this point-in-time scan does not prevent a subsequently rolled-back/pre-feature PD from becoming leader. That old leader omits `default-store-limit` on every full-config persist (including unrelated config updates), so the configured default can still be silently lost during rollback even though newer leaders would reject further default changes.

Comment (King-Dylan) [reply server/util.go]: Good point. The member scan is only an activation guard, so it is insufficient for preserving the value across a later rollback.

I will update the PR to also store a compatibility copy in `schedule.store-limit[0]` (store ID 0 is invalid for a real store). New versions will keep `default-store-limit` and entry 0 synchronized in the same persisted JSON. A pre-feature PD already understands and rewrites the `store-limit` map, so an unrelated full-config persist preserves entry 0; after re-upgrade, a new PD can restore `default-store-limit` from it. The reserved entry will be filtered from public config/store-limit APIs and metrics. This follows the same general pattern as #9134: reuse an old persisted representation with a reserved compatibility identifier.

The existing scan will remain as the mixed-version activation guard, while entry 0 provides rollback durability. I will also add a regression test for new write -> pre-feature full-config rewrite -> new reload, including asymmetric and explicit-zero defaults. An unpatched old PD still will not apply entry 0 to newly added stores after restart; strict behavioral equivalence while running the downgraded version would require a separate release-branch backport.

File: pkg/mcs/scheduling/server/apis/v1/api.go [hunk 5]
```diff
@@ -362,6 +362,7 @@ func getConfig(c *gin.Context) {
 		c.String(http.StatusInternalServerError, "failed to decode primary's config: "+err.Error())
 		return
 	}
+	primaryCfg.Schedule.MigrateDeprecatedFlags()
```
Comment (rleungx) [inline pkg/mcs/scheduling/server/apis/v1/api.go]: When this handler serves a follower, `MigrateDeprecatedFlags` inserts the reserved store-limit entry 0, and the merged config is returned without removing it. Consequently, `/scheduling/api/v1/config` exposes a phantom store limit only on followers, making the public config response depend on which scheduling member receives the request.

Comment (King-Dylan) [reply pkg/mcs/scheduling/server/apis/v1/api.go]: Good catch. Fixed in 08e8cb58d. The follower still runs MigrateDeprecatedFlags to normalize a config returned by an older primary, then applies CloneWithoutDefaultStoreLimitCompat before merging it into the public response. I also added a unit regression for the legacy payload and integration assertions that both primary and follower /config responses omit store-limit[0] and return the same default-store-limit.

File: pkg/mcs/discovery/registry_entry.go [hunk 6]
```diff
@@ -32,6 +32,8 @@ type ServiceRegistryEntry struct {
 	GitHash        string `json:"git-hash"`
 	DeployPath     string `json:"deploy-path"`
 	StartTimestamp int64  `json:"start-timestamp"`
+	// Features maps a supported feature to the build hash that advertised it.
+	Features map[string]string `json:"features,omitempty"`
```
Comment (rleungx) [inline pkg/mcs/discovery/registry_entry.go]: Is there any way that we can avoid introducing it?

Comment (King-Dylan) [reply pkg/mcs/discovery/registry_entry.go]: Yes. Reworked this in 7a2441321 to avoid extending ServiceRegistryEntry, and removed the PD member feature keys as well.

Before changing the default, PD now requests each registered PD and Scheduling Service member's existing config endpoint and checks the raw response for schedule.default-store-limit. The request is forced to be handled and serialized by the target member, so a pre-feature binary omits the field and blocks activation without requiring a version threshold or new registry metadata.

The store-limit[0] rollback compatibility remains unchanged. I added coverage for supported, pre-feature, malformed, and unavailable responses, multiple PD client URLs, and an independent Scheduling Service registry containing a pre-feature member.

File: server/config/persist_options.go [hunk 7]
```diff
@@ -83,6 +83,7 @@ func (o *PersistOptions) GetScheduleConfig() *sc.ScheduleConfig {
 
 // SetScheduleConfig sets the PD scheduling configuration.
 func (o *PersistOptions) SetScheduleConfig(cfg *sc.ScheduleConfig) {
+	cfg.SyncDefaultStoreLimitCompat()
```
Comment (rleungx) [inline server/config/persist_options.go]: When schedule persistence fails, `Server.SetScheduleConfig` passes its previously published `old` config back here, and this call writes `StoreLimit[0]` into that shared map before the atomic store. Concurrent readers can still hold and iterate the old config, so this failure path introduces a data race and can terminate PD with a concurrent map read/write fatal error.

File: server/handler.go [hunk 8]
```diff
@@ -255,6 +255,9 @@ func (h *Handler) SetAllStoresLimit(ratePerMin float64, limitType storelimit.Typ
 	if err != nil {
 		return err
 	}
+	if err := h.s.checkDefaultStoreLimitPersistenceSupport(); err != nil {
```
Comment (rleungx) [inline server/handler.go]: When `ratePerMin` already equals the persisted default for `limitType`, this operation only needs to normalize existing per-store entries, but the unconditional probe still rejects it if any registered member is old or temporarily unavailable. This blocks a safe existing-store update during rolling upgrades even though the cluster-level default does not change.

Comment (King-Dylan) [reply server/handler.go]: Fixed in f25a5669a. SetAllStoresLimit now compares the requested rate with the persisted default for the requested limit type and runs the compatibility probe only when that default would actually change. A same-default request can still normalize existing per-store entries while an old Scheduling Service member is registered; the integration test verifies that case succeeds while a real default change is still rejected.

File: server/config/persist_options.go (hunk 7 above)
Comment (King-Dylan) [reply server/config/persist_options.go]: Fixed in f25a5669a. PersistOptions.SetScheduleConfig now clones the caller-owned schedule config before synchronizing the store-limit[0] compatibility entry and publishing it. The persistence rollback therefore no longer writes to the previously published map. I applied the same ownership rule to the Scheduling Service PersistConfig setter and added regression tests that verify the input remains untouched; the affected packages also pass under go test -race.

File: server/cluster/cluster.go [hunk 9]
```diff
@@ -2389,10 +2389,7 @@ func (c *RaftCluster) AddStoreLimit(store *metapb.Store) {
 		return
 	}
 
-	slc := sc.StoreLimitConfig{
-		AddPeer:    sc.DefaultStoreLimit.GetDefaultStoreLimit(storelimit.AddPeer),
-		RemovePeer: sc.DefaultStoreLimit.GetDefaultStoreLimit(storelimit.RemovePeer),
-	}
+	slc := cfg.GetDefaultStoreLimit()
```
Comment (rleungx) [inline server/cluster/cluster.go]: `AddStoreLimit` can retain a stale schedule snapshot while a concurrent `/stores/limit` request publishes and persists a new default. If the store registration then reaches `SetScheduleConfig` last, it overwrites the successful all-store update—including `DefaultStoreLimit`, the compatibility entry, and existing per-store limits—and persists the new store with the old default.

File: pkg/schedule/config/config.go [hunk 10]
```diff
@@ -353,6 +393,23 @@ func (c *ScheduleConfig) Clone() *ScheduleConfig {
 	return &cfg
 }
 
+// SyncDefaultStoreLimitCompat stores the default in a representation that
+// pre-feature PD versions already understand and preserve.
+func (c *ScheduleConfig) SyncDefaultStoreLimitCompat() {
+	if c.StoreLimit == nil {
+		c.StoreLimit = make(map[uint64]StoreLimitConfig)
+	}
+	c.StoreLimit[DefaultStoreLimitCompatStoreID] = c.DefaultStoreLimit
```
Comment (rleungx) [inline pkg/schedule/config/config.go]: `store-limit[0]` preserves the configured default across an old-version rewrite, but a pre-feature PD does not use it when registering a store. If such a PD becomes leader after rollback, it persists the built-in value as that store's higher-priority per-store entry, so the incorrect limit survives the subsequent re-upgrade even though the cluster default is restored.

File: server/cluster/cluster.go (hunk 9 above)
Comment (King-Dylan) [reply server/cluster/cluster.go]: Fixed in 8465b4577. All store-limit configuration writes (`AddStoreLimit`, `RemoveStoreLimit`, `SetStoreLimit`, and `SetAllStoresLimit`) are now serialized by a dedicated lock, so a stale registration snapshot cannot publish or persist after a concurrent update. Added `TestStoreLimitUpdatesAreSerialized`, which blocks the registration persist and verifies the all-store update waits and then updates both the persisted default and the new store entry.

File: pkg/schedule/config/config.go (hunk 10 above)
Comment (King-Dylan) [reply pkg/schedule/config/config.go]: Agreed. This means `store-limit[0]` only provided partial rollback preservation and could not make a pre-feature downgrade safe without a release-branch backport. We have narrowed the compatibility contract instead: rolling upgrade is supported, but downgrade to a pre-feature PD or Scheduling Service after changing the persisted default is explicitly unsupported. Commit 8465b4577 removes the reserved entry and all restoration/filtering logic, keeps the member capability probe as the rolling-upgrade activation gate, and documents the persistent per-store consequence in the PR description and release note.

File: server/util.go (hunk 4 above)
Comment (King-Dylan) [reply server/util.go]: Follow-up after the later downgrade finding: we no longer claim rollback durability. Commit 8465b4577 removes `store-limit[0]`; the PR now supports mixed-version rolling upgrade through the activation gate and explicitly treats downgrade to a pre-feature binary after changing the persisted default as unsupported.

File: pkg/mcs/scheduling/server/apis/v1/api.go (hunk 5 above)
Comment (King-Dylan) [reply pkg/mcs/scheduling/server/apis/v1/api.go]: The reserved compatibility entry has now been removed entirely in 8465b4577 because downgrade to a pre-feature binary is outside the supported contract. Therefore this follower-only phantom entry and its filtering path no longer exist.

File: server/cluster/cluster.go [hunk 11]
```diff
@@ -2590,15 +2597,17 @@ func (c *RaftCluster) SetStoreLimit(storeID uint64, typ storelimit.Type, ratePer
 
 // SetAllStoresLimit sets all store limit for a given type and rate.
 func (c *RaftCluster) SetAllStoresLimit(typ storelimit.Type, ratePerMin float64) error {
+	c.storeLimitLock.Lock()
```
Comment (rleungx) [inline server/cluster/cluster.go]: Because `storeLimitLock` only covers the RaftCluster store-limit helpers, another whole-config writer such as `runStoreConfigSync` or `Server.SetScheduleConfig` can snapshot the old schedule before this call, then persist that snapshot after this call succeeds. The acknowledged default and per-store limits are then reverted in storage (and by MCS watchers), so a restart or leadership change can still give future stores the old limit.

File: server/cluster/cluster_test.go [hunk 12]
```diff
@@ -3344,6 +3344,118 @@ func TestStoreLimitChangeRefreshLimiter(t *testing.T) {
...
+
+type blockingSaveConfigStorage struct {
+	storage.Storage
+	saveStarted chan struct{}
+	releaseSave chan struct{}
+	once        sync.Once
+}
+
+func (s *blockingSaveConfigStorage) SaveConfig(cfg any) error {
+	s.once.Do(func() {
```
Comment (rleungx) [inline server/cluster/cluster_test.go]: `sync.Once.Do` blocks every concurrent caller until the first invocation returns, so the second `SaveConfig` remains blocked here even if `storeLimitLock` is removed. Since the final assertions inspect only the in-memory config, this test passes without the serialization it is intended to verify and cannot catch the stale-persistence regression.

File: server/cluster/cluster.go (hunk 11 above)
Comment (King-Dylan) [reply server/cluster/cluster.go]: Fixed in feb2026de. I moved the serialization boundary from the RaftCluster-only `storeLimitLock` to `PersistOptions`. `Persist` and the new schedule read-modify-publish-persist transaction now share `persistMu`, and the persisted snapshot is constructed only after that lock is acquired. The store-limit helpers use this transaction, so `runStoreConfigSync` or another full-config persistence cannot write an older schedule after an acknowledged store-limit update. The full schedule-config HTTP paths also pass the snapshot they were built from and reject the update with a retry error if it has changed. The regression test covers a generic full-config persistence racing with `SetAllStoresLimit` and reloads the final value from storage.

File: server/cluster/cluster_test.go (hunk 12 above)
Comment (King-Dylan) [reply server/cluster/cluster_test.go]: Fixed in feb2026de. The test storage no longer uses `sync.Once.Do`; it counts calls under a mutex and blocks only the first `SaveConfig`, so a second save would proceed immediately if production code did not serialize it. The test starts a generic full-config `Persist`, races `SetAllStoresLimit`, verifies the latter cannot finish before the first save is released, and then reloads the config from storage to verify the persisted default. It therefore fails when the production serialization is removed and no longer relies only on the in-memory value.

File: server/config/persist_options.go [hunk 13]
```diff
@@ -783,12 +826,20 @@ type persistedConfig struct {
...
-	return o.Persist(storage)
+	return o.UpdateScheduleConfig(storage, func(_ *sc.ScheduleConfig, next *sc.ScheduleConfig) (bool, error) {
+		next.StoreLimitVersion = storelimit.VersionV2
+		return true, nil
+	})
 }
 
 // Persist saves the configuration to the storage.
 func (o *PersistOptions) Persist(storage endpoint.ConfigStorage) error {
+	o.persistMu.Lock()
```
Comment (rleungx) [inline server/config/persist_options.go:863]: This mutex only starts after callers have already published their component updates. Schedule writers such as `Coordinator.InitSchedulers` still clone the full `ScheduleConfig`, later call `SetScheduleConfig`, and only then call `Persist`. If `SetAllStoresLimit` commits while scheduler initialization is using its old clone, the initializer can publish that stale clone before entering this lock and then persist it after the successful store-limit update, reverting the new default (and other concurrent schedule changes) in memory and storage. The new serialization therefore fixes unrelated full-config writers, but the remaining schedule read-modify-persist paths can still overwrite the acknowledged update.

Comment (King-Dylan) [reply server/config/persist_options.go]: Fixed in 0bb1ff6e0. You are right that serializing only Persist was insufficient because scheduler writers could publish a stale full ScheduleConfig before entering it. InitSchedulers now commits only its Schedulers field through SchedulerConfigProvider.UpdateScheduleConfig, so the clone used for scheduler discovery cannot replace concurrent fields. PD and Scheduling Service runtime schedule field helpers now also read-modify-publish under the same schedule transaction; dynamic AddSchedulerCfg/RemoveSchedulerCfg no longer publish an unlocked full clone. TestInitSchedulersPreservesConcurrentScheduleUpdate blocks initialization after its initial clone, commits SetAllStoresLimit, resumes init, and verifies both memory and reloaded storage retain the new default.

File: pkg/schedule/config/config.go [hunk 14]
```diff
@@ -227,6 +229,11 @@ type ScheduleConfig struct {
 	// StoreBalanceRate is the maximum of balance rate for each store.
 	// WARN: StoreBalanceRate is deprecated.
 	StoreBalanceRate float64 `toml:"store-balance-rate" json:"store-balance-rate,omitempty"`
+	// DefaultStoreLimit is the default limit of scheduling for stores.
+	DefaultStoreLimit StoreLimitConfig `toml:"default-store-limit" json:"default-store-limit"`
+	// defaultStoreLimitJSONPresence distinguishes omitted fields in persisted legacy JSON
+	// from explicitly configured zero values.
+	defaultStoreLimitJSONPresence *storeLimitConfigJSONPresence
```
Comment (rleungx) [inline pkg/schedule/config/config.go]: `defaultStoreLimitJSONPresence` is decoder metadata rather than schedule configuration, and keeping it on the runtime `ScheduleConfig` creates unnecessary hidden state. `UnmarshalJSON` also runs for partial `/config/schedule` requests; that path does not call `MigrateDeprecatedFlags`, so the marker can remain in the atomic config, be copied by `Clone`, and participate in `reflect.DeepEqual`. Omitted-vs-explicit-zero detection belongs to the persisted/remote decoding boundary (for example, pointer or raw input fields passed into migration), leaving the runtime config with only behavioral values.

File: server/handler.go [hunk 15]
```diff
@@ -255,6 +255,19 @@ func (h *Handler) SetAllStoresLimit(ratePerMin float64, limitType storelimit.Typ
...
+	defaultStoreLimit := c.GetScheduleConfig().GetDefaultStoreLimit()
+	defaultChanged := true
+	switch limitType {
+	case storelimit.AddPeer:
+		defaultChanged = ratePerMin != defaultStoreLimit.AddPeer
+	case storelimit.RemovePeer:
+		defaultChanged = ratePerMin != defaultStoreLimit.RemovePeer
+	}
+	if defaultChanged {
+		if err := h.s.checkDefaultStoreLimitPersistenceSupport(); err != nil {
```
Comment (rleungx) [inline server/handler.go]: `/stores/limit` previously updated existing stores without depending on every PD or Scheduling Service member. When the requested rate changes the default, this preflight now makes that existing operation fail if any registered member is pre-feature or temporarily unreachable, even though updating the existing per-store limits remains valid. Scoping the persisted-default guarantee to a fully upgraded cluster preserves the old API behavior without adding a synchronous all-member availability dependency.

File: server/api/config.go [hunk 16]
```diff
@@ -473,7 +475,7 @@ func (h *confHandler) SetScheduleConfig(w http.ResponseWriter, r *http.Request)
 		return
 	}
 
-	if err := h.svr.SetScheduleConfig(*config); err != nil {
+	if err := h.svr.SetScheduleConfigIfUnchanged(*oldConfig, *newConfig); err != nil {
```
Comment (rleungx) [inline server/api/config.go]: `/config/schedule` is a partial update, but this path materializes a full snapshot and later compares the entire `ScheduleConfig`. A concurrent change to an unrelated schedule field therefore turns an otherwise mergeable request into an error and requires the extra exported `SetScheduleConfigIfUnchanged`/`CompareAndPersistScheduleConfig` layer. Applying the JSON patch to the latest clone inside the existing schedule update transaction preserves concurrent fields without full-snapshot CAS.

File: pkg/schedule/config/config_provider.go [hunk 17]
```diff
@@ -69,6 +69,10 @@ type SchedulerConfigProvider interface {
 
 	GetScheduleConfig() *ScheduleConfig
 	SetScheduleConfig(*ScheduleConfig)
+	UpdateScheduleConfig(
```
Comment (rleungx) [inline pkg/schedule/config/config_provider.go]: `UpdateScheduleConfig` exposes PD's persistence transaction through the shared scheduler-provider interface, but the Scheduling Service implementation has to ignore `storage` because its config is watcher-owned. The only cross-provider production caller added here is `Coordinator.InitSchedulers`, which updates just the scheduler list. A focused scheduler-list update operation would keep persistence ownership in each implementation and avoid expanding this store-limit change into a generic callback API plus mechanical rewrites of unrelated or test-only setters.

File: server/handler.go (hunk 15 above)
Comment (King-Dylan) [reply server/handler.go]: Agreed and fixed in 5c414acf7. The synchronous PD and Scheduling Service member probe and its gate tests are removed, so `/stores/limit` keeps its previous availability even when a registered Scheduling Service member is old or unreachable. The compatibility contract is now scoped explicitly: existing per-store updates work during rolling upgrade, while the persisted default for future stores is guaranteed after every PD and Scheduling Service member supports the field; if changed during the mixed-version window, reapply it after upgrade. The PR description and release note now state this. The attached E2E report still validates the core leader-transfer and restart behavior; its gate section was simulated, not a real mixed-binary deployment.

File: server/api/config.go (hunk 16 above)
Comment (King-Dylan) [reply server/api/config.go]: Fixed in 5c414acf7. `/config/schedule` now applies its raw JSON patch to the latest clone inside `UpdateScheduleConfig`; the legacy `/config` single-item path does the same via `SetScheduleConfigItem`. The full-snapshot CAS APIs are removed. `TestConcurrentPartialScheduleConfigUpdatesDoNotConflict` runs concurrent partial schedule and default-store-limit updates, requires every update to succeed, and verifies memory matches reloaded storage. This directly removes the six CAS-conflict 500s observed in the attached E2E report at 0bb1ff6e.

File: pkg/schedule/config/config.go (hunk 14 above)
Comment (King-Dylan) [reply pkg/schedule/config/config.go]: Fixed in 5c414acf7. `ScheduleConfig` no longer has custom JSON decoder state: `defaultStoreLimitJSONPresence` and `UnmarshalJSON` are removed. Persisted, watched, and remote full-config decoders now pass the raw schedule JSON to `MigrateDeprecatedFlagsFromJSON`, which consumes omitted-vs-explicit-zero presence at that boundary and leaves only behavioral fields in the runtime config. Partial `/config/schedule` decoding therefore cannot leave hidden state in the atomic config or affect `Clone` or `DeepEqual`. Legacy reload, explicit-zero, remote PD, and MCS watcher tests pass.

File: pkg/schedule/config/config_provider.go (hunk 17 above)
Comment (King-Dylan) [reply pkg/schedule/config/config_provider.go]: Fixed in 5c414acf7. The shared provider no longer exposes the generic callback. It now has only `SetSchedulers(SchedulerConfigs)`. `InitSchedulers` calls that focused merge and then each provider implementation uses its existing `Persist`: PD persists the latest config under `persistMu`, while Scheduling Service keeps its watcher-owned no-op persistence. The storage argument is no longer leaked into the Scheduling Service implementation, and the unrelated and test-only setter rewrites from 0bb1ff6e were reverted. The scheduler-init concurrency regression still passes.

**Review by rleungx — APPROVED** [review]

**Comment by rleungx** [comment]
This PR adds a new config and needs @niubell to approve.

**Review by niubell — APPROVED** [review]
